### PR TITLE
cpu/memory manager containerMap memory leak

### PR DESCRIPTION
**Pull request by Dingshujie**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
if cpu manager/ memory manager policy set to none, no one remove container info from containerMap, finaly lead memory leak.

this pr call cpu manager/memory manager RemoveContainer in PostStopContainer, to remove  container info from containerMap
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
/assign

@klueska as well.

**Comment by klueska** [comment]
We cannot do what you propose here. We used to do what you are doing but it causes issues when containers are restarted (within a still running pod). We need to maintain the containerMap across these restarts.

we may indeed have a memory leak for the case of the none policy, but we will need to solve it another way.

**Comment by Dingshujie** [comment]
> We cannot do what you propose here. We used to do what you are doing, but it causes issues when containers are restarted (within a still running pod). We need to maintain the `containerMap` across these restarts.
> 
> We may indeed have a memory leak for the case of the `none` policy, but we will need to solve it another way.

got restart scene, i will try another method to solve it

**Comment by Dingshujie** [comment]
@klueska find another method to slove this leak,  Can you take a look at this PR again?   thanks a lot.

File: pkg/kubelet/cm/containermap/container_map.go [hunk 1]
```diff
@@ -69,3 +71,19 @@ func (cm ContainerMap) GetContainerRef(containerID string) (string, string, erro
 	}
 	return cm[containerID].podUID, cm[containerID].containerName, nil
 }
+
+// ListContainerRef lists all container ref pair from container map,
+func (cm ContainerMap) ListContainerRef() map[string]sets.String {
```
Comment (klueska) [inline pkg/kubelet/cm/containermap/container_map.go:76]: I'm not a fan of this function that returns an unintuitive data structure like this. If we want to add something that walks all of the entries in the container map, I'd be fine with that though. Something like:
```
func (cm ContainerMap) Visit(f func(podUID, containerName, containerID string)) {
	for k, v := range cm {
		f(v.podUID, v.containerName, k)
	}
}
```
Which (following your code below) could then be called as:
```
cm.Visit(func(podUID, containerName, containerID string) {
	if _, ok := activeContainers[podUID][containerName]; !ok {
		klog.ErrorS(nil, "RemoveStaleState: removing container", "podUID", podUID, "containerName", containerName)
		err := m.policyRemoveContainerByRef(podUID, containerName)
		if err != nil {
			klog.ErrorS(err, "RemoveStaleState: failed to remove container", "podUID", podUID, "containerName", containerName)
		}
	}
})
```

Comment (Dingshujie) [reply pkg/kubelet/cm/containermap/container_map.go]: appreciate. change to visit, PTAL @klueska

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 2]
```diff
@@ -388,6 +388,16 @@ func (m *manager) removeStaleState() {
...
+
+	m.containerMap.Visit(func(podUID, containerName, containerID string) {
+		if _, ok := activeContainers[podUID][containerName]; !ok {
+			klog.ErrorS(nil, "RemoveStaleState: removing container", "podUID", podUID, "containerName", containerName)
+			err := m.policyRemoveContainerByRef(podUID, containerName)
+			if err != nil {
+				klog.ErrorS(err, "RemoveStaleState: failed to remove container", "podUID", podUID, "containerName", containerName)
+			}
+		}
+	})
```
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:400]: LGTM.

And just for reference, removing from the map while iterating through it IS supported by the golang specification:

> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next. If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced. If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.

**Review by klueska — COMMENTED** [review]
/lgtm
/approve

**Comment by rphillips** [comment]
It appears we have to wait for the tree to thaw for 1.25 for this fix. That thaw will happen on or around April 19.

**Comment by Dingshujie** [comment]
Can we add this to v1.24 milestone? @rphillips @liggitt

**Comment by liggitt** [comment]
> It appears we have to wait for the tree to thaw for 1.25 for this fix

Bugfixes can be merged during code freeze if impact/risk is justified.

Is this a regression? (looks like not, since there are cherry-picks open)

What is the risk of the fix? Is it changing feature-gated or optionally executed code? If so, what feature gate or opt-in configuration exposes the bug?

**Comment by liggitt** [comment]
I also wouldn't expect us to take a change without accompanying tests

**Comment by Dingshujie** [comment]
> I also wouldn't expect us to take a change without accompanying tests

**Comment by Dingshujie** [comment]
sorry， careless close pr.

**Comment by Dingshujie** [comment]
> > It appears we have to wait for the tree to thaw for 1.25 for this fix
> 
> Bugfixes can be merged during code freeze if impact/risk is justified.
> 
> Is this a regression? (looks like not, since there are cherry-picks open)
> 
> What is the risk of the fix? Is it changing feature-gated or optionally executed code? If so, what feature gate or opt-in configuration exposes the bug?

cpu manager default policy is none, if pod repeated create and delete, kubelet memory will keep growing.
also cpu manager  policy set to static， none exclusive pod container id also will leak in containermaps,  kubelet memory will keep growing.

**Comment by rphillips** [comment]
@Dingshujie Since the project is in a frozen state, the criteria is does the PR fix a regression within the 1.24 release. Waiting for the thaw minimizes churn/bugs/regressions added during the release process.

**Comment by Dingshujie** [comment]
> @Dingshujie Since the project is in a frozen state, the criteria is does the PR fix a regression within the 1.24 release. Waiting for the thaw minimizes churn/bugs/regressions added during the release process.

ok, let's wait for 1.24 released.

**Review by yanghesong — APPROVED** [review]

**Comment by pacoxu** [comment]
This will be included in 1.25. So a cherry-pick PR for 1.24 was needed as well. (It is ok to open one after v1.24 is released.)

**Pull request merged**